File: crates/nil-client/src/client/world.rs
// Copyright (C) Tsukilabs contributors
// SPDX-License-Identifier: AGPL-3.0-only

use super::Client;
use crate::error::Result;
use nil_core::world::{WorldConfig, WorldStats};
use std::path::PathBuf;

impl Client {
  /// GET `/world/config`
  pub async fn get_world_config(&self) -> Result<WorldConfig> {
    self.http.get_json("world/config").await
  }

  /// GET `/world/stats`
  pub async fn get_world_stats(&self) -> Result<WorldStats> {
    self.http.get_json("world/stats").await
  }

  /// POST `/world/save`
  pub async fn save_world(&self, path: PathBuf) -> Result<()> {
    self.http.post("world/save", path).await
  }
}

File: app/src/commands/world.ts
// Copyright (C) Call of Nil contributors
// SPDX-License-Identifier: AGPL-3.0-only

import { invoke } from '@tauri-apps/api/core';
import type { SavedataInfo } from '@/core/savedata';
import { WorldConfigImpl } from '@/core/model/world-config';
import { type RawWorldStats, WorldStatsImpl } from '@/core/model/stats/world-stats';

export async function getWorldConfig(): Promise<WorldConfigImpl> {
  const config = await invoke<WorldConfig>('get_world_config');
  return WorldConfigImpl.create(config);
}

export async function getWorldStats(): Promise<WorldStatsImpl> {
  const stats = await invoke<RawWorldStats>('get_world_stats');
  return WorldStatsImpl.fromRaw(stats);
}

export async function readSavedataInfo(path: string) {
  return invoke<SavedataInfo>('read_savedata_info', { path });
}

export async function saveWorld(path: string) {
  await invoke('save_world', { req: { path } });
}
